src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,274 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.Json;
+
+namespace NuGet.ProjectModel
+{
+    /// <summary>
+    /// This struct is used to read over a memeory stream in parts, in order to avoid reading the entire stream into memory.
+    /// It functions as a wrapper around <see cref="Utf8JsonStreamReader"/>, while maintaining a stream and a buffer to read from.
+    /// </summary>
+    internal ref struct Utf8JsonStreamReader
+    {
+        private static readonly char[] DelimitedStringDelimiters = new char[] { ' ', ',' };
+        private const int BufferSizeDefault = 16 * 1024;
+        private ReadOnlySpan<byte> _utf8Bom = new byte[] { 0xEF, 0xBB, 0xBF };
+        private Utf8JsonReader _reader;
+        // The buffer is used to read from the stream in chunks.
+        private byte[] _buffer;
+        private bool _disposed;
+
+        internal Utf8JsonStreamReader(Stream stream) : this(stream, ArrayPool<byte>.Shared.Rent(BufferSizeDefault))
+        {
+
+        }
+
+        internal Utf8JsonStreamReader(Stream stream, byte[] buffer)
+        {
+            if (stream is null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            _disposed = false;
+            Stream = stream;
+            _buffer = buffer;
+            Stream.Read(_buffer, 0, 3);
+            var offset = 0;
+            if (!_utf8Bom.SequenceEqual(_buffer.AsSpan(0, 3)))
+            {
+                offset = 3;
+            }
+            var blocksRead = Stream.Read(_buffer, offset, _buffer.Length - offset);
+
+            _reader = new Utf8JsonReader(_buffer.AsSpan(0, blocksRead + offset), isFinalBlock: blocksRead + offset < _buffer.Length, state: new JsonReaderState(new JsonReaderOptions
```

Comment:
As mentioned elsewhere, Stream.Read doesn't promise that the buffer will be filled, only that at least 1 byte will be filled. Therefore for certain `Stream` implementations, the `isFinalBlock` calculation will be incorrect.

Also, consider using variables for the method parameters. I'm not a fan of expressions in method parameters, but I don't remember seeing other people in the team give similar PR feedback, so maybe I'm the only one who cares.  If I can't hover my mouse in the debugger to see the exact value being passed to the method (so `someObject.PropertyName` is fine), I'd like a variable, so I can inspect the value when debugging.